cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll

```diff
@@ -47,11 +48,60 @@ class PureFunction extends ArrayFunction, TaintFunction {
       )
     ) and
     (
+      output.isOutReturnValue() or
+      output.isOutReturnPointer()
+    )
+  }
+
+  override predicate parameterNeverEscapes(int i) {
+    getParameter(i).getUnspecifiedType() instanceof PointerType and
+    not (
+      i = 0 and
+      getType().getUnspecifiedType() instanceof PointerType
+    )
+  }
+
+  override predicate parameterEscapesOnlyViaReturn(int i) {
+    i = 0 and
+    getType().getUnspecifiedType() instanceof PointerType
+  }
+
+  override predicate parameterIsAlwaysReturned(int i) {
+    none()
+  }
+
+  override predicate neverReadsMemory() {
+    none()
+  }
+  
+  override predicate neverWritesMemory() {
+    any()
+  }
+}
+
+class PureFunction extends TaintFunction, SideEffectFunction {
+  PureFunction() {
+    exists(string name |
+      hasName(name) and
```

Comment:
I know it's independent of this PR, but this use of `hasName` has bitten us several times, and I don't want it to bite us again. It's just very likely that class member functions have short names like `abs`. See, for example, #1023.

Please switch to `hasGlobalName` here and in `PureStrFunction`. If you want to make sure that `std::abs` is also matched, then it's time to implement the `hasGlobalOrStdName` predicate we've been talking about.

Reply:
I'll switch to `hasGlobalName` for now, and do `hasGlobalOrStdName` as a separate PR (since it should involve a pass through most of the library)